core/ir/riscv64/instr.c

```diff
@@ -0,0 +1,451 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL RIVOS, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "../globals.h"
+#include "instr.h"
+
+bool
+instr_set_isa_mode(instr_t *instr, dr_isa_mode_t mode)
+{
+    return (mode == DR_ISA_RV64IMAFDC);
+}
+
+dr_isa_mode_t
+instr_get_isa_mode(instr_t *instr)
+{
+    return DR_ISA_RV64IMAFDC;
+}
+
+int
+instr_length_arch(dcontext_t *dcontext, instr_t *instr)
+{
+    /* FIXME i#3544: C has shorter instruction length. */
+    return RISCV64_INSTR_SIZE;
+}
+
+bool
+opc_is_not_a_real_memory_load(int opc)
+{
+    ASSERT_NOT_IMPLEMENTED(false);
+    return opc == OP_l;
+}
+
+uint
+instr_branch_type(instr_t *cti_instr)
+{
+    /* FIXME i#3544: Add parameter checking (inc table 2.1 in spec) */
+    int opcode = instr_get_opcode(cti_instr);
+    switch (opcode) {
+    case OP_jal: return LINK_DIRECT | LINK_CALL;
+    case OP_jalr: return LINK_INDIRECT | LINK_CALL;
+    case OP_beq:
+    case OP_bne:
+    case OP_blt:
+    case OP_bltu:
+    case OP_bge:
+    case OP_bgeu: return LINK_DIRECT | LINK_JMP;
+    }
+    CLIENT_ASSERT(false, "instr_branch_type: unknown opcode");
+    return LINK_INDIRECT;
+}
+
+const char *
+get_opcode_name(int opc)
+{
+    ASSERT_NOT_IMPLEMENTED(false);
+    return "";
+}
+
+bool
+instr_is_mov(instr_t *instr)
+{
+    ASSERT_NOT_IMPLEMENTED(false);
+    return false;
+}
+
+bool
+instr_is_call_arch(instr_t *instr)
+{
+    ASSERT_NOT_IMPLEMENTED(false);
+    return false;
+}
+
+bool
+instr_is_call_direct(instr_t *instr)
+{
+    /* FIXME i#3544: Check if valid */
+    int opc = instr_get_opcode(instr);
+    return (opc == OP_jal);
+}
+
+bool
+instr_is_near_call_direct(instr_t *instr)
+{
+    /* FIXME i#3544: Check if valid */
+    int opc = instr_get_opcode(instr);
+    return (opc == OP_jal);
+}
+
+bool
+instr_is_call_indirect(instr_t *instr)
+{
+    /* FIXME i#3544: Check if valid */
+    int opc = instr_get_opcode(instr);
+    return (opc == OP_jalr);
+}
+
+bool
+instr_is_return(instr_t *instr)
+{
+    /* FIXME i#3544: Check if valid */
+    int opc = instr_get_opcode(instr);
+    opnd_t rd = instr_get_dst(instr, 0);
+    opnd_t rs = instr_get_src(instr, 0);
+    return (opc == OP_jalr && opnd_get_reg(rd) == DR_REG_X0 &&
+            opnd_is_memory_reference(rs) && opnd_get_base(rs) == DR_REG_X1 &&
+            opnd_get_disp(rs) == 0);
+}
+
+bool
+instr_is_cbr_arch(instr_t *instr)
+{
+    ASSERT_NOT_IMPLEMENTED(false);
```

Comment:
Since instr_invert_cbr is implemented, makes sense to implement this one too.

Reply:
For that matter I should also implement `instr_is_ubr_arch` which is trivial.